### PR TITLE
UNR-443: Troublshooting docs ++

### PR DESCRIPTION
**Pull request by danielimprobable**
#### Description
 This PR contains a set of different documents that have been brought over to MD. 

#### Documentation
https://improbableio.atlassian.net/browse/UNR-443

#### Primary reviewers
@anne-edwards @ElleEss @improbable-valentyn

### CODE REVIEW

File: docs/content/troubleshooting.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+**Q:** I’m getting the error `Unrecognized type 'UCommander' - type must be a UCLASS, USTRUCT or UENUM` when building my project in Visual Studio.
+
+**A:** This is most likely due to not having run the legacy SDK code generation. Try running `Scripts/Codegen.bat` at least once. This should generate the required files for continuing the build process.
+
+------
+
+**Q:** When I try to start SpatialOS from the toolbar plugin, I get the following notification:
+`[improbable.worker.assembly.WorkerAssemblyProvider] The worker assembly does not contain any worker configurations. No workers will be able to connect to this deployment. Unless you specifically intended to start a deployment without any workers, please make sure your assembly was built correctly, and in case of a cloud deployment, also make sure it was uploaded correctly.`
+
+**A:** This is an indication that you haven’t built the worker configurations for the server and the clients. You can fix this by running the `Scripts/BuildWorkerConfig.bat` script. This generates the worker configs which allow your workers to connect to the local instance of SpatialOS.
```
Comment (anne-edwards) [inline docs/content/troubleshooting.md]: Can we change "server" and "clients" to "server worker" and "client workers" here? @danielimprobable 
I would have made the change directly but I wasn't sure if "server worker" needed to be singular, and "client workers" plural, or if they could both be plural.

Comment (danielimprobable) [reply docs/content/troubleshooting.md]: I thought we went away form calling them server worker and client workers as the worker wasn't needed, I think giray corrected that in the original doc. I think we should use plurals for both

File: docs/known-issues.md [hunk 2]
```diff
@@ -1 +1,25 @@
...
+## Unreal GDK Known Issues
+
+Known issue = any major user-facing bug or lack of user-facing feature that:
+1. diverges from vanilla Unreal design or implementation (e.g. ordering of reliable RPCs), **OR**
+1. diverges from user expectations from a SpatialOS project (e.g. interacting across worker boundaries)
+
+| Issue | Date added | Ticket | Workaround? | Done? |
+|-------|-------------------|--------|-------------|-------|
+| Order of reliable RPCs is not respected in case of unresolved UObject* parameter. | 2018-06-14 | [UNR-363](https://improbableio.atlassian.net/browse/UNR-336) | Potential user-code workaround. | No|
+| Reliable multicast RPCs aren’t supported. | 2018-06-14 | | Ask yourself if you really need a reliable multicast RPC. | No |
```
Comment (anne-edwards) [inline docs/known-issues.md]: "As yourself if..." - this sounds like we're not planning to fix the issue, in which case I don't think it's fair to include it as a known issue.

Comment (anne-edwards) [reply docs/known-issues.md]: Also, I think it sounds a little bit rude - maybe rephrase, offering an alternative to using reliable multicast RPCs?

Comment (danielimprobable) [reply docs/known-issues.md]: Yes, it is more of an unsupported issue, we cover it in the trouble shooting guide, maybe we can remove it

File: docs/known-issues.md [hunk 3]
```diff
@@ -1 +1,25 @@
...
\ No newline at end of file
+## Unreal GDK Known Issues
+
+Known issue = any major user-facing bug or lack of user-facing feature that:
+1. diverges from vanilla Unreal design or implementation (e.g. ordering of reliable RPCs), **OR**
+1. diverges from user expectations from a SpatialOS project (e.g. interacting across worker boundaries)
+
+| Issue | Date added | Ticket | Workaround? | Done? |
+|-------|-------------------|--------|-------------|-------|
+| Order of reliable RPCs is not respected in case of unresolved UObject* parameter. | 2018-06-14 | [UNR-363](https://improbableio.atlassian.net/browse/UNR-336) | Potential user-code workaround. | No|
```
Comment (anne-edwards) [inline docs/known-issues.md]: Can we say anything more about what the user-code workaround might look like?

File: docs/known-issues.md [hunk 4]
```diff
@@ -1 +1,25 @@
...
+| Reliable multicast RPCs aren’t supported. | 2018-06-14 | | Ask yourself if you really need a reliable multicast RPC. | No |
+| Project build `PCHUsageMode` must be set to `UseExplicitOrSharedPCHs`. | 2018-06-27 |[UNR-354](https://improbableio.atlassian.net/browse/UNR-354)| | No |
+| Adaptive Unity builds are not supported. | 2018-07-28 | | Disable `bUseAdaptiveUnityBuild` following [Epic’s documentation](https://docs.unrealengine.com/en-US/Programming/UnrealBuildSystem/Configuration). | No |
+| When working on a blueprint class, we need to specify the blueprint class in the DefaultEditorSpatialGDk.ini. E.g SampleGameCharacterBP, but we actually need to add SampleGameCharacterBP_C, the _C is an Unreal workaround. | 2018-06-27 | [UNR-363](https://improbableio.atlassian.net/browse/UNR-363) | Append `_C` to blueprint classes names in the `.ini` file. | No |
```
Comment (anne-edwards) [inline docs/known-issues.md]: I don't understand this - could you clarify whether "we" should say "you" (as in, the user)? 

And the bit that says "but we actually need to..." is a bit confusing - it looks like we're saying you need to do one thing, and then that you actually need to do a different thing.

EDIT: I think I've just got it - would this be better?

> When working on a Blueprint class, we need to specify the Blueprint class in the `DefaultEditorSpatialGDk.ini` and add `_C` onto the end of it - for example, `SampleGameCharacterBP_C`. (The `_C` is an Unreal workaround.)

File: docs/known-issues.md (hunk 3 above)
Comment (danielimprobable) [reply docs/known-issues.md]: @improbable-valentyn do you have any notes on what the workaround was for this issue? have the user wait for the successfull execution of the unresolved RPC before calling the second resolved one?

File: docs/known-issues.md [hunk 5]
```diff
@@ -1 +1,25 @@
...
+| Dynamic component replication. | 2018-07-03 | [UNR-366](https://improbableio.atlassian.net/browse/UNR-366) | Mimic behavior with static components (not really a workaround). | No |
```
Comment (anne-edwards) [inline docs/known-issues.md]: The workaround here says "not really a workaround" 😂 
We should probably remove this or add a better workaround if possible!

File: docs/known-issues.md [hunk 6]
```diff
@@ -1 +1,25 @@
...
+| Components with replicated sub-objects fail. | 2018-07-03 |[UNR-417](https://improbableio.atlassian.net/browse/UNR-417) | Make all components top-level. | No |
+| Multiple replicated components of the same type. | 2018-07-03 | [UNR-416](https://improbableio.atlassian.net/browse/UNR-416) | | No |
```
Comment (anne-edwards) [inline docs/known-issues.md]: What about them? From the JIRA ticket it looks like something fails, but I'm not sure what.

File: docs/known-issues.md (hunk 4 above)
Comment (danielimprobable) [reply docs/known-issues.md]: As it turns out, this is no longer an issue, so we should remove the entry

File: docs/known-issues.md (hunk 5 above)
Comment (danielimprobable) [reply docs/known-issues.md]: I would remove the "not really a workaround"

File: docs/known-issues.md (hunk 6 above)
Comment (danielimprobable) [reply docs/known-issues.md]: You cant have more than one component of the same type on your actor at the moment, you will get an error when you try to generate interop in such a scenario

File: docs/known-issues.md [hunk 7]
```diff
@@ -1 +1,25 @@
...
+|Interop codegen fails when the destination file is locked. The workflow is less than ideal as you can't run it again until you restart the Unreal Editor. | 2018-07-16 | [UNR-350](https://improbableio.atlassian.net/browse/UNR-350) | Ensure destination files/folders are unlocked. | No |
+| Deleting entities directly from the Inspector causes undefined behavior. | 2018-07-17 | [UNR-425](https://improbableio.atlassian.net/projects/UNR/issues/UNR-425) | | No |
+| Interop code generation stalls if its child processes produce console output during execution. |2018-07-19 | [UNR-422](https://improbableio.atlassian.net/browse/UNR-422) | Disable verbose output. This stops interop code generation from stalling, but you won't receive error messages if it fails. | No |
+| Having generated files in the symlinked folder `Source/SpatialGDK` means you can’t work in two projects at the same time using the same clone of the SpatialOS Unreal GDK. | 2018-07-23 | [UNR-458](https://improbableio.atlassian.net/browse/UNR-458) | Separate UnrealGDK checkout for each project. | No |
```
Comment (anne-edwards) [inline docs/known-issues.md]: I changed "checkout" to "clone" - is that correct? Just realised I also need to do it for the workaround.

File: docs/known-issues.md [hunk 8]
```diff
@@ -1 +1,25 @@
...
+| We don't support listen servers. | 2018-07-30 | | Use dedicated servers instead. | No |
```
Comment (anne-edwards) [inline docs/known-issues.md]: I updated this because I googled "listen servers" and "dedicated servers" and it looks like they shouldn't be capitalised.

File: docs/known-issues.md [hunk 9]
```diff
@@ -1 +1,25 @@
...
+| Compilation error for RPCs that take structs with private members. | 2018-07-16 | [UNR-144](https://improbableio.atlassian.net/browse/UNR-144) | Add the type binding as a friend class. | No |
+|Interop codegen fails when the destination file is locked. The workflow is less than ideal as you can't run it again until you restart the Unreal Editor. | 2018-07-16 | [UNR-350](https://improbableio.atlassian.net/browse/UNR-350) | Ensure destination files/folders are unlocked. | No |
+| Deleting entities directly from the Inspector causes undefined behavior. | 2018-07-17 | [UNR-425](https://improbableio.atlassian.net/projects/UNR/issues/UNR-425) | | No |
+| Interop code generation stalls if its child processes produce console output during execution. |2018-07-19 | [UNR-422](https://improbableio.atlassian.net/browse/UNR-422) | Disable verbose output. This stops interop code generation from stalling, but you won't receive error messages if it fails. | No |
```
Comment (anne-edwards) [inline docs/known-issues.md]: I updated this one a bit - please check it's still correct.

File: docs/known-issues.md (hunk 7 above)
Comment (danielimprobable) [reply docs/known-issues.md]: yep that is correct

File: docs/known-issues.md (hunk 9 above)
Comment (danielimprobable) [reply docs/known-issues.md]: Sounds good to me:)

File: docs/content/troubleshooting.md (hunk 1 above)
Comment (anne-edwards) [reply docs/content/troubleshooting.md]: I think @ElleEss had second thoughts about this and we decided to go for "server-worker" and "client-worker" (yes, including the hyphen, which I'm not convinced by!). Just to be completely explicit. So I'll update these for now, and we can try to push the rest of the company to follow suit.

File: docs/known-issues.md (hunk 3 above)
Comment (improbable-valentyn) [reply docs/known-issues.md]: I guess they would have to enforce the order explicitly in gameplay as you suggested (instead of relying on the RPCs being executed in order).

Comment (anne-edwards) [reply docs/known-issues.md]: Cool - how about "Potential user-code workaround to explicitly enforce the order of RPCs"?

**Review by anne-edwards — APPROVED** [review]
I think this is now good to go!

File: docs/known-issues.md (hunk 9 above)
Comment (m-samiec) [reply docs/known-issues.md]: We've already removed the workaround, so if users hit this issue it won't apply. Worth removing reference the workaround for now.

File: docs/content/porting-unreal-project-to-gdk.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+* [Prerequisites](#prerequisites)
+* [Setting up the project structure](#setting-up-the-project-structure)
+* [Adding the SpatialGDK module to your project](#adding-the-spatialgdk-module-to-your-project)
+* [Modifying Unreal classes for GDK compatibility](#modifying-unreal-classes-for-gdk-compatibility)
+* [Adding Unreal GDK configurations](#adding-unreal-gdk-configurations)
+* [Generating code for SpatialOS C++ workers and building your project](#generating-code-for-spatialos-c++-workers-and-building-your-project)
+* [Running the Interop Code Generator and setting up the UE4 Editor](#running-the-interop-code-generator-and-setting-up-the-UE4-editor)
+* [Running your game](#running-your-game)
+
+
```
Comment (Vatyx) [inline docs/content/porting-unreal-project-to-gdk.md]: Extra newline?

**Comment by Vatyx** [comment]
@danielimprobable What do you think about adding an issue with not being able to find a `UClass` or something similar?
